feat: add ansible task testing infrastructure based on Docker and pytest

jfroche opened the pull request:
This complements the existing AMI tests in testinfra by providing a faster feedback loops for Ansible development without requiring a full VM.

We are also using testinfra to validate that the Ansible tasks have the desired effect.

It is based on Docker, it can be run locally (e.g. macOS) or in CI.

Note that this approach is not intended to replace the AMI tests, but rather to provide a more efficient way to test Ansible tasks during development.

You can run the tests using `nix run -L .\#ansible-test`

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * CI workflow to build/publish multi-architecture test images and run the Ansible test suite.

* **Tests**
  * Added Ansible Nginx playbook and pytest tests to validate nginx automatically.
  * Added a local/CI test runner to execute the test suite.

* **Quality / Tooling**
  * Added Ansible lint configuration and integrated an ansible-lint pre-commit hook.
  * Added packaged test images and Nix helpers for building/testing.

* **Bug Fixes / Chores**
  * Fixed a YAML comment parse issue and normalized task titles/modes for consistency.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jfroche committed:
feat: add ansible task testing infrastructure based on Docker and pytest

This complements the existing AMI tests in testinfra by providing
a faster feedback loops for Ansible development without requiring a full
VM.

We are also using testinfra to validate that the Ansible tasks have the
desired effect.

It is based on Docker, it can be run locally (e.g. macOS) or in CI.

Note that this approach is not intended to replace the AMI tests, but
rather to provide a more efficient way to test Ansible tasks during
development.

You can run the tests using `nix run -L .\#ansible-test`

hunleyd commented on .github/actions/nix-install-ephemeral/action.yml:
```diff
@@ -0,0 +1,46 @@
+    - name: Install nix
+      uses: cachix/install-nix-action@v31
+      with:
+        install_url: https://releases.nixos.org/nix/nix-2.31.2/install
```
AIUI, @samrose wants us to use Nix 2.29.x for now for $reasons

jfroche replied:
Good catch I will upgrade in another PR

hunleyd commented on ansible/tests/nginx.yaml:
```diff
@@ -0,0 +1,14 @@
+- hosts: localhost
+  tasks:
+  - name: Install dependencies
+    apt:
```
```suggestion
    ansible.builtin.apt:
```

jfroche replied:
Thank you for the reminder, I wanted to automate the ansible-lint run on all our ansible files.
I have started with that little test and its dependencies here: https://github.com/supabase/postgres/pull/1882/commits/7388b0bfae0d6b0d8d9ad356fb7fa1dee84f6d2c

We could go further and fix/validate all the other task files in another PR. Also there might be some rules you are not interested to follow ?

hunleyd replied:
suggested ansible-lint rc:
```
---
profile: production

# exclude_paths included in this file are parsed relative to this file's location
# and not relative to the CWD of execution. CLI arguments passed to the --exclude
# option are parsed relative to the CWD of execution.
exclude_paths:
  - .cache/  # implicit unless exclude_paths is defined in config
  - .github/

# Mock modules or roles in order to pass ansible-playbook --syntax-check

# mock_modules:

# mock_roles:

use_default_rules: true
enable_list:
  - args
  - empty-string-compare
  - no-log-password
  - no-same-owner
warn_list:
  - experimental
skip_list:
  - name[casing]
  - name[prefix]
  - yaml[line-length]
  - var-naming[no-role-prefix]

# Offline mode disables installation of requirements.yml
offline: false

# Make the output more readable
parseable: true

# Define required Ansible's variables to satisfy syntax check
# extra_vars:

# List of additional kind:pattern to be added at the top of the default
# match list, first match determines the file kind.
kinds:
  - meta: "**/meta/main.yml"
  - playbook: "**/playbooks/*.playbook.yml"
  - tasks: "**/tasks/*.yml"
  - vars: "**/vars/*.yml"
```

might need some tweaking for paths, but this is what we used at Crunchy and it seemed to work well

jfroche replied:
Based on your file I have added a similar configuration. We aren't validating all the tasks in `ansible/tasks` in the CI but we could do this in another PR.

hunleyd commented on ansible/tests/nginx.yaml:
```diff
@@ -0,0 +1,14 @@
+    apt:
+      pkg:
+        - build-essential
+      update_cache: yes
```
```suggestion
      update_cache: true
```

hunleyd commented on ansible/tests/nginx.yaml:
```diff
@@ -0,0 +1,14 @@
+      pkg:
+        - build-essential
+      update_cache: yes
+  - import_tasks: ../tasks/setup-nginx.yml
```
```suggestion
  - ansible.builtin.import_tasks:
      file: ../tasks/setup-nginx.yml
```

hunleyd commented on ansible/tests/nginx.yaml:
```diff
@@ -0,0 +1,14 @@
+    service:
+      name: nginx
+      state: started
+      enabled: yes
```
```suggestion
    ansible.builtin.service:
      enabled: true
      name: 'nginx'
      state: 'started'
```

hunleyd reviewed (changes requested):
mostly ansible-lint fixups

jfroche committed:
feat: add ansible-lint validation for test playbooks

Configure pre-commit hook to run ansible-lint on test playbooks and their dependencies.
Since test playbooks include tasks from existing task files, ansible-lint automatically validates those dependencies as well.

hunleyd commented on ansible/tasks/setup-nginx.yml:
```diff
@@ -1,78 +1,82 @@
   ansible.builtin.user:
-    name: 'nginx'
-    state: 'present'
+    name: nginx
```
quote dropping is intentional?

samrose replied:
probably not is my guess. maybe needs rebase

jfroche replied:
I ran `ansible-lint --fix` and it automatically removes unnecessary quotes to follow YAML best practices for improved readability (https://docs.ansible.com/projects/lint/rules/yaml/#yaml).

hunleyd replied:
interesting. it seems they updated https://redhat-cop.github.io/automation-good-practices/#_yaml_and_jinja2_syntax since my last reading

samrose commented:
@hunleyd I realize  it may not yet be documented very well, but in order for you to get 1:1 parity with the linter in this PR, you would want to:

1. check out the branch test/ansible-tasks
2. run `nix develop -c ansible-lint <the rest of your command here>`

By doing it this way, you'll then use the same ansible lint version in this PR, and the same configuration. 

This helps avoid the "works on my machine" problem that may come up if we install ansible lint and tools from brew install etc. We can get 1:1 with what other developers are doing with `nix develop -c ansible-lint <the rest of your command here>` plus what the CI job is doing which will also use the exactly same version and setup. I hope that helps, and thanks as ever for reviewing these PR.

jfroche commented:
Also we have pre-commit hook in the dev shell. So if you start the dev shell using `nix develop`, pre-commit hook will be register and triggered when you commit/push but you can also run it manually inside the devshell (just like the CI will do): `pre-commit run --all`..

The pre-commit hook configuration is defined in `nix/hooks.nix` and as you will see we only validates github action, ansible-lint and formatting (defined in `nix/fmt.nix`) for the moment.

hunleyd commented on .github/actionlint.yaml:
perhaps i'm just missing something, but why have .github/actionlint.yaml *and* .github/actionlint.yml ? or at least, why not symlink one to the other?

jfroche replied:
Good point, merge problem here too. Thank you !

jfroche committed:
feat: run ansible tests in parallel with pytest-xdist

samrose commented:
@mmlb this one might be useful for future work, so leaving it open for us to review
